### ISSUE TITLE
我已解决test_dice极低的问题 | I have resolved the problem that the test_dice  is extremely low

### ISSUE THREAD

**Issue by LambdaLi** [lee-zq/3DUNet-Pytorch#39]
![4219d3454a33e3d8670b9c5962935a49](https://github.com/lee-zq/3DUNet-Pytorch/assets/131449535/c672128f-1cb7-48e4-9787-59c0d3e3e784)

![d7ed6caf4221afffec5cf90fbb07ef42_720](https://github.com/lee-zq/3DUNet-Pytorch/assets/131449535/fb7788b6-2402-4b3b-9618-a9c032a50f07)

这个代码我跑了四次，最后发现是自己对题主的意思理解错误导致，然后我花了几天时间逐行阅读代码，希望我的遭遇能够帮助到各位：`

①不要预处理‘测试集(test)’！！！不要预处理‘测试集(test)’！！！不要预处理‘测试集(test)’！！！重要的事情说三遍，只需要预处理训练集(train)，dataset_lits_test.py会对测试(test)集图像进行处理。

②前面前辈说到的ResUNet.py里135-161行的map里的upsample里的scale参数后两位减半。

③把test.py第25行的注释给取消注释，该行的作用看注释的注释，做完后便！！！不！！！需要像前辈之前说的运行test.py时把②里说到的减半参数->进行恢复。
-------------------------------------------------------------------------------------------------------------------------------------------------
I ran this code four times and finally realized that it was due to my incorrect understanding of the author's meaning. I spent a few days reading the code line by line, hoping that my experience can help everyone:
① Do not preprocess the 'test set'!!! Do not preprocess the 'test set'!!! Do not preprocess the 'test set'!!! Say important things three times, just preprocess the training set, dataset_ Lits_ Test. py will process the test set images.
② The last two digits of the scale parameter in the upsample in the map from lines 135 to 161 of ResUNet. py mentioned by the previous generation are halved.
③ Uncomment the 25th line of test.py. The purpose of this line depends on the annotation. Once it is done, it will be ready.We don't need to restore the parameters halved mentioned in ② when running test. py.

**Comment by LambdaLi** [comment]
![d7ed6caf4221afffec5cf90fbb07ef42_720](https://github.com/lee-zq/3DUNet-Pytorch/assets/131449535/a98b17d9-786f-4d38-a35b-4ba313c8b17d)

**Comment by zz10001** [comment]
厉害了👍

**Comment by zhouyizhuo** [comment]
@LambdaLi @zz10001 您好，我想问一下`slice_down_scale`的含义具体是什么。我现在想将此网络应用于新的数据集，如果我更改`min_slices`和`slice_down_scale`的值，我需要更改模型的其他参数吗？
![image](https://github.com/lee-zq/3DUNet-Pytorch/assets/89564922/fb3a0e30-abad-4c53-b56b-1685ffb12d53)

**Comment by LambdaLi** [comment]
> @LambdaLi @zz10001 您好，我想问一下`slice_down_scale`的含义具体是什么。我现在想将此网络应用于新的数据集，如果我更改`min_slices`和`slice_down_scale`的值，我需要更改模型的其他参数吗？ ![image](https://private-user-images.githubusercontent.com/89564922/287608032-fb3a0e30-abad-4c53-b56b-1685ffb12d53.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE2NzIxNTQsIm5iZiI6MTcwMTY3MTg1NCwicGF0aCI6Ii84OTU2NDkyMi8yODc2MDgwMzItZmIzYTBlMzAtYWJhZC00YzUzLWI1NmItMTY4NWZmYjEyZDUzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA0VDA2MzczNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3ODEzNmE5ZjVjOGI2NDg5MGU1NTc0MjYzMTBlNWY4ZDI1ZTFhZWRlNzY5MDZhN2UzZmNmNzUyYjJmZWM3NDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FiwN-j4CT8L3Xv0yqC9Up8OqJriPHPlep9XrXpXaA5s)

slice_down_scale这个是处理CT的Z轴切片，它和上面xy_down_scale作用应该是一样，我个人理解的是可以在x, y轴上缩放(下采样)来降低图片尺寸，Z轴由于是3D分割所以维持原状，毕竟要找到肝脏的首部和尾部，改变之后影不影响分割性能我也不太清楚，min_slices我看了相关文章"To maximize the use of GPU video memory, 48 consecutive slices were randomly selected from these as network model inputs with a size of 48×256×256."，这个是每次送进去多少切片，具体怎么改自己斟酌一下吧，有误的地方还请谅解。

**Comment by zhouyizhuo** [comment]
好的，非常感谢您的回复。还想再问一下，在训练的时候没有问题，但是在val处理某个文件时，报了以下错误，请问可能会是什么原因？
![image](https://github.com/lee-zq/3DUNet-Pytorch/assets/89564922/6be6192b-147a-4d2f-9ba0-cbeec94c2826)
![image](https://github.com/lee-zq/3DUNet-Pytorch/assets/89564922/c283d71c-7e7e-4ba7-84e8-221266d818b6)

其余的没有问题
![image](https://github.com/lee-zq/3DUNet-Pytorch/assets/89564922/31ae782f-8253-4d98-8990-963a36c90488)

**Comment by LambdaLi** [comment]
> 好的，非常感谢您的回复。还想再问一下，在训练的时候没有问题，但是在val处理某个文件时，报了以下错误，请问可能会是什么原因？ ![image](https://private-user-images.githubusercontent.com/89564922/287618499-6be6192b-147a-4d2f-9ba0-cbeec94c2826.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE2NzQ4MTAsIm5iZiI6MTcwMTY3NDUxMCwicGF0aCI6Ii84OTU2NDkyMi8yODc2MTg0OTktNmJlNjE5MmItMTQ3YS00ZDJmLTliYTAtY2JlZWM5NGMyODI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA0VDA3MjE1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmODU0ODBlZDE5NTI4Njk0ZTcyNTRhOWI3NDBhOGQ5ZmRhZGVlMzJkZjk4ZGJkNTg3YjRjMGZjODFmZjMxMGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.O1IYIM7eEMQZ2jkFnxQA6JIWSGdJdVFGn7XWx_iLi-Q) ![image](https://private-user-images.githubusercontent.com/89564922/287618825-c283d71c-7e7e-4ba7-84e8-221266d818b6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE2NzQ4MTAsIm5iZiI6MTcwMTY3NDUxMCwicGF0aCI6Ii84OTU2NDkyMi8yODc2MTg4MjUtYzI4M2Q3MWMtN2U3ZS00YmE3LTg0ZTgtMjIxMjY2ZDgxOGI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA0VDA3MjE1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM5Nzg5YjRiMjJmOTM1NDJjN2Q0OTI5ZTJlYzcyMTg3YTkxOWZmZDNhY2UyMWIxNmIzNTUzNjJhMmI1MzJmMmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.r7wlQPpSST-i-VEz2RjXgJoTAeRyfvC9LoeIGs9KapY)
> 
> 其余的没有问题 ![image](https://private-user-images.githubusercontent.com/89564922/287618620-31ae782f-8253-4d98-8990-963a36c90488.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE2NzQ4MTAsIm5iZiI6MTcwMTY3NDUxMCwicGF0aCI6Ii84OTU2NDkyMi8yODc2MTg2MjAtMzFhZTc4MmYtODI1My00ZDk4LTg5OTAtOTYzYTM2YzkwNDg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA0VDA3MjE1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ0YmRjNmQ3ZDg4MzhmOTM4ZDYwM2FhY2Y3MzI5MzVlNGFmOTg0ZDg5YWNhN2FlYTU4OTM0ZTUwNDM5MjJhZDEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.c_O1KxdsLny7aUAr4slv6MRjt9nsDQehvy58LkRrfck)

突然出问题确实很奇怪，分辨率也太不正常检查一下 下采和上采吧

**Comment by zz10001** [comment]
> 好的，非常感谢您的回复。还想再问一下，在训练的时候没有问题，但是在val处理某个文件时，报了以下错误，请问可能会是什么原因？

不好意思这份代码太久没看了有点忘记了，只是之前参考了另一份类似的代码处理规则；顺便推荐下可以参考我空间2d的处理，希望对你有帮助

**Comment by zhouyizhuo** [comment]
@LambdaLi 您好，感谢您的帮助，我已经解决了无法开始训练的问题。目前我遇到了训练中突然loss变得很高的问题，请问这是正常的吗？顺便我想问一下， 我的训练集有250个左右，我应该设置多少个epoch才比较合适？
![image](https://github.com/lee-zq/3DUNet-Pytorch/assets/89564922/44988df1-d969-4765-999c-31355384f6d6)

**Comment by LambdaLi** [comment]
@zhouyizhuo 很不正常，我跑了几次这个模型loss是持续降低的，而且你看你两个dice低的离谱，由于这个只是我练手用的，所以我没有关闭他的早停功能，3080ti跑了60左右的epoch就停了，A4000跑了100多个epoch吧，但是我觉得问题不是出在epoch上，还是检查代码，是不是模型预测的结果和真实标签的不对应。

**Comment by RY-97** [comment]
> 好的，非常感谢您的回复。还想再问一下，在训练的时候没有问题，但是在val处理某个文件时，报了以下错误，请问可能会是什么原因？ ![image](https://private-user-images.githubusercontent.com/89564922/287618499-6be6192b-147a-4d2f-9ba0-cbeec94c2826.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzMTAwMTcsIm5iZiI6MTcxMjMwOTcxNywicGF0aCI6Ii84OTU2NDkyMi8yODc2MTg0OTktNmJlNjE5MmItMTQ3YS00ZDJmLTliYTAtY2JlZWM5NGMyODI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA1VDA5MzUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzYjlkMmUwNzE1YmUxZGMzYjA2YWEwZDAxZTlhMWM1M2JiNTFhNjM0OTg4OWNjNDAyOWNlNjI5NTZmZGU3MTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xeIDsZX6UOWayN-Dt5QHuDjhAJJArpJ1ShBN8PF0O_M) ![image](https://private-user-images.githubusercontent.com/89564922/287618825-c283d71c-7e7e-4ba7-84e8-221266d818b6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzMTAwMTcsIm5iZiI6MTcxMjMwOTcxNywicGF0aCI6Ii84OTU2NDkyMi8yODc2MTg4MjUtYzI4M2Q3MWMtN2U3ZS00YmE3LTg0ZTgtMjIxMjY2ZDgxOGI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA1VDA5MzUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExMDczNDBjY2MyOWQ4OWYwNWU1OGYzZWQzODMwN2IyOTA0YjMwYmE5MGY3YjkwZjA1ZDNiZTVjNGUxNGVhNTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0ERq6ovbr5Uz_0lBg4OAUEfcCr8DIoMb7e9Ox0WQcUI)
> 
> 其余的没有问题 ![image](https://private-user-images.githubusercontent.com/89564922/287618620-31ae782f-8253-4d98-8990-963a36c90488.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTIzMTAwMTcsIm5iZiI6MTcxMjMwOTcxNywicGF0aCI6Ii84OTU2NDkyMi8yODc2MTg2MjAtMzFhZTc4MmYtODI1My00ZDk4LTg5OTAtOTYzYTM2YzkwNDg4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDA1VDA5MzUxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI2YmZlMmZhNzE2ZDFlZTllMzg3MmRiZjBmNzhmNDAxNjU4NmQ5N2M5YzQ3YzJjZjViYWZjODJjYmM2ZDczMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.YtQGOishGG1xtwVt3RwldmmuQmpepGeh6uvYsdig5O8)

请问你可以给我你的代码一份？

**Comment by RY-97** [comment]
> @zhouyizhuo 很不正常，我跑了几次这个模型loss是持续降低的，而且你看你两个dice低的离谱，由于这个只是我练手用的，所以我没有关闭他的早停功能，3080ti跑了60左右的epoch就停了，A4000跑了100多个epoch吧，但是我觉得问题不是出在epoch上，还是检查代码，是不是模型预测的结果和真实标签的不对应。

可以共享一份代码吗？
